### PR TITLE
Add letsencrypt and gcsfuse roles

### PR DESCRIPTION
**Pull request by mrgnr**
This PR provides the ability to generate TLS certificates for nginx using the [webroot plugin](https://letsencrypt.readthedocs.io/en/latest/using.html#webroot) for Let's Encrypt. The webroot plugin uses the [http-01](https://github.com/ietf-wg-acme/acme/blob/master/draft-ietf-acme-acme.md#user-content-http) challenge and stores challenge response data in the directory specified by `letsencrypt_webroot`.

When generating certificates behind a load balancer, it's necessary for all servers in the pool to have access to the challenge response data. On GCP, [gcsfuse](https://github.com/GoogleCloudPlatform/gcsfuse) can be used to mount a shared Google Storage bucket to the `letsencrypt_webroot` directory.

Example configuration:

```
# Directory where the challenge response data will be stored
letsencrypt_webroot: "/var/www/letsencrypt"

# User that needs access to `letsencrypt_webroot` (i.e. the user that runs nginx)
letsencrypt_webuser: "www-data"

# This will generate two separate certs, one for foo.example.com and
# bar.example.com, and the other for baz.example.com
letsencrypt_certs:
 - domains: ["foo.example.com", "bar.example.com"]
 - domains: ["baz.example.com"]

# Set this if your sites-available directory is in a weird place
letsencrypt_webserver_sites_available: "/edx/app/nginx/sites-available"

# Email address associated with generated certificates
letsencrypt_email: "info@example.com"

# Extra options to pass to letsencrypt
# You can use `--staging` to generate non-valid certs for testing
letsencrypt_flags: "--staging"

# Google Storage buckets and their mount points
gcsfuse_mounts:
  - bucket: "my-bucket"
    path: "{{ letsencrypt_webroot }}"
    owner: "{{ letsencrypt_webuser }}"
    group: "{{ letsencrypt_webuser }}"
    mode: "0770"
    keyfile: "/etc/credentials/creds.json"

# base64-encoded contents of json credentials file for a service account with
# access to the bucket
secret_gcsfuse_keyfile: |
    Ym/20U0uqaLgtsMKHAG6TbGoM6Y9B27ruX0rSXb6F4dRIgoTbJ20Lg+e/UxQSPBWaTsh6QXl9noI
    ew3BumVLglS14ws4okuMHtPCycx33G/uEdSM5U/wRjOaBNiWV6SM/O2NQg==

gcsfuse_keyfiles:
  - content: "{{ secret_gcsfuse_keyfile }}"
    dest: "/etc/credentials/creds.json"
    owner: "{{ letsencrypt_webuser }}"
    group: "{{ letsencrypt_webuser }}"
    mode: "0440"
```

### CODE REVIEW

**Comment by mrgnr** [comment]
[(8) Ficus: Generate TLS certs with Let's Encrypt](https://trello.com/c/0cuX97vd/919-8-ficus-generate-tls-certs-with-let-s-encrypt)

**Comment by johnbaldwin** [comment]
@mrgnr +1 for the great PR description!

File: gcsfuse/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    content: "{{ item.content | b64decode }}"
+    dest: "{{ item.dest }}"
+    owner: "{{ item.owner | default(omit) }}"
+    group: "{{ item.group | default(omit) }}"
+    mode: "{{ item.mode | default(omit) }}"
+  with_items: "{{ gcsfuse_keyfiles }}"
+  tags: ['gcsfuse', 'gcsfuse:configuration']
+
+- name: Mount buckets in directories
+  shell: gcsfuse {{ (item.keyfile != None) | ternary("--key-file " + item.keyfile, "") }} "{{ item.bucket }}" "{{ item.path }}" && touch "{{ item.path }}/.mounted"
```
Comment (johnbaldwin) [inline gcsfuse/tasks/main.yml]: I'm not asking to change this, but curious, would breaking this into multiple lines break it?

File: letsencrypt/templates/nginx/letsencrypt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{% set domains = [] %}
```
Comment (johnbaldwin) [inline letsencrypt/templates/nginx/letsencrypt]: Do we want to use the .j2 extension as a convention?

**Comment by johnbaldwin** [comment]
@mrgnr Just a couple of small comments/questions.  Looking forward to having this be part of our workflow!

File: letsencrypt/templates/nginx/letsencrypt (hunk 2 above)
Comment (mrgnr) [inline letsencrypt/templates/nginx/letsencrypt]: I don't have strong feelings, but I don't really see a need to suffix everything in `templates/` with .j2. It's mostly an edX convention, and I don't find it helpful tbh.

File: gcsfuse/tasks/main.yml (hunk 1 above)
Comment (mrgnr) [inline gcsfuse/tasks/main.yml]: @johnbaldwin It is possible. I can give you something like

```
  shell: >
    gcsfuse {{ (item.keyfile != None) | ternary("--key-file " + item.keyfile, "") }} "{{ item.bucket }}" "{{ item.path }}" &&
    touch "{{ item.path }}/.mounted"
```
if you'd prefer.

File: letsencrypt/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+---
+
+letsencrypt_webroot: "/var/www/letsencrypt"
+letsencrypt_webuser: "www-data"
+letsencrypt_webserver: "nginx"
+letsencrypt_webserver_config: "nginx/letsencrypt"
+letsencrypt_webserver_sites_available: "/etc/nginx/sites-available"
+letsencrypt_webserver_sites_enabled: "/etc/nginx/sites-enabled"
+letsencrypt_https_redirect: false
+letsencrypt_email: !!null
```
Comment (denibertovic) [inline letsencrypt/defaults/main.yml]: This should default to ops@appsembler.com I think....Unless the point is that these roles are OSS...then yeah maybe not.

File: letsencrypt/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    letsencrypt certonly {{ letsencrypt_flags }} --agree-tos -m "{{ letsencrypt_email }}"
+    --webroot -w "{{ letsencrypt_webroot }}" -d {{ item.domains | join(' -d ') }}
+  args:
+    creates: /etc/letsencrypt/live/{{ item.domains[0] }}/fullchain.pem
+  with_items: "{{ letsencrypt_certs }}"
+  become: True
+  become_user: "{{ letsencrypt_webuser }}"
+  tags: ['letsencrypt']
+
+- name: Create cron job to renew certificates
```
Comment (denibertovic) [inline letsencrypt/tasks/main.yml]: Is this necessary? From what I know systemd installs a task automatically once you install certbot via apt.

File: letsencrypt/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
+---
+
+- name: Install Let's Encrypt
+  apt: name="letsencrypt"
```
Comment (denibertovic) [inline letsencrypt/tasks/main.yml]: why not use the certbot name?

File: letsencrypt/defaults/main.yml (hunk 3 above)
Comment (mrgnr) [inline letsencrypt/defaults/main.yml]: Yep, I set this in our Ficus base vars: https://github.com/noderabbit-team/edx-configs/pull/75/files
With this being a public repo, I'd prefer not to get notices about other peoples' expired certs :p

File: letsencrypt/tasks/main.yml (hunk 5 above)
Comment (mrgnr) [inline letsencrypt/tasks/main.yml]: The package name and command vary by platform, but it's `letsencrypt` on Xenial. I'll make this a variable so it's generic.

File: letsencrypt/tasks/main.yml (hunk 4 above)
Comment (mrgnr) [inline letsencrypt/tasks/main.yml]: It doesn't as far as I know.. I've always had to create a cron job. Here's what comes with the package:

```
# dpkg-query -L letsencrypt
/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/letsencrypt
/usr/share/doc/letsencrypt/README.rst.gz
/usr/share/doc/letsencrypt/copyright
/usr/share/man
/usr/share/man/man1
/usr/share/man/man1/letsencrypt.1.gz
/usr/bin
/usr/bin/letsencrypt
/usr/share/doc/letsencrypt/changelog.Debian.gz
```

**Pull request merged**